Add skeleton for wizard steps.

aaronfc opened the pull request:
Fixes #5126

### Changes proposed in this Pull Request
* Created `Wizard` component that expects a `steps` attribute containing an ordered array of the different `Steps` that should be rendered.
* Some sample `Steps` have been added in order to test the different cases and flows we expect to support:
  * `course-details-step.js`: Has a sample input and way to persist data in the `Wizard`'s data. It has also a sample action that displays the contents of the `Wizard` data.
  * `upgrade-step.js`: Empty step with only two actions: next and go to some url.
  * `course-patterns-step.js`: Empty step with only a final action - `onCompletion` can be called when choosing the pattern at some point.
  * `lesson-details-step.js` and `lesson-patterns-step.js` are also empty steps that are not currently used. 

### Testing instructions
- Create a new Lesson or Course and you will be prompted with a sample **wizard for Course**.
- Play around with the actions.
- Check the code :)

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video
https://user-images.githubusercontent.com/799065/169290375-2e84138d-aa37-4a16-a457-4dd1fb33f025.mp4

### Discussion / Context
- This is only a proposal for the basic structure for the Wizard.
- Tried by using `{children}` but couldn't figure out the best way to have different "actions" per step - if anyone wants to give it a try I will be pleased to learn how to do that!
- I expect this task to include some minor CSS for the modal, but specific details for each step will be handled in each steps task.
- I will look how to have different steps for different post types but I want to get fast feedback on the main structure. Will update the PR with that once I have it.

### TODO
- [x] Allow different steps depending on post type.
- [x] CSS for modal + wizard structure.
- [x] Update modal title? Maybe title can be part of each step.
- [x] Add tests for `wizard.js`.

aaronfc committed:
Add skeleton for wizard steps.

- Added `Wizard` component which expects a list of `Step` components.
- Some sample `Step` components have been added to `steps/`.
- Each Step can also define a `.Actions` component to define the
  actions.
- Some minor css has been added but must be considered something
  temporary.
- Alternative implementation using `{children}` got a little bit complex
  regarding the step actions.

renatho commented on assets/admin/editor-wizard/wizard.js:
```diff
@@ -0,0 +1,53 @@
+
+	const goToNextStep = () => {
+		if ( currentStepNumber + 1 < steps.length ) {
+			setCurrentStepNumber( currentStepNumber + 1 );
```
Probably not critical here, but just a tip for the future: Always when a new state depends on the previous one, it's nice to set it with a callback getting the previous state, like this:

```
setCurrentStepNumber( prev => prev + 1 );
```

It helps to avoid async weird issues when using the setter function quickly.

aaronfc replied:
Cool! I had to do this in the Standalone Questions too but I forgot. Thanks for refreshing me about it!

Edit: I was going to change it but I think I will keep it as it is. I want the "if" condition and the set be consistent. I think using `prev => prev + 1` could have some side-effect regarding the limits. What do you think?

renatho replied:
Makes sense! And this case won't be a big deal because it's something very unlike to call quickly in a row.

fjorgemota replied:
>  I want the "if" condition and the set be consistent. 

In this case, one possibility would be to have the `if` inside the callback. React is smart enough to not trigger a change/re-render if the value returned by the callback is the same as the previously stored, as [mentioned here](https://reactjs.org/docs/hooks-reference.html#usestate):

> If your update function returns the exact same value as the current state, the subsequent rerender will be skipped completely.

The main issue with this approach would be the call to the `onCompletion` callback, which would be definitely weird to call here. Maybe it would make sense to call it in a `useEffect` hook or similar, tho, based on this same condition. :thinking: 

> And this case won't be a big deal because it's something very unlike to call quickly in a row.

While I agree with this, at the same time I fear double-clicks. I don't know about you, but as an advanced user, I'm very optimized to click NEXT -> NEXT -> NEXT -> FINISH very quickly just to finish these wizards quickly. Do we have some protection in that case to avoid this?

WDYT @renatho @aaronfc?

aaronfc replied:
> The main issue with this approach would be the call to the onCompletion callback, which would be definitely weird to call here. Maybe it would make sense to call it in a useEffect hook or similar, tho, based on this same condition. 🤔

This made sense to me, so I was implementing it when I realised this would imply setting a "currentStep" out of range, which complicates a little the logic. So I will keep it as it is. I think the current risk would be that some "clicks" might be ignored, and that's only for REALLY fast clicks. So I think we are safe :)

renatho commented on assets/admin/editor-wizard/wizard.js:
```diff
@@ -0,0 +1,53 @@
+ * @param {Array}    props.steps        Array with the steps that will be rendered.
+ * @param {Function} props.onCompletion Callback to call when wizard is completed.
+ */
+const Wizard = ( { steps, onCompletion } ) => {
```
I liked how you prepared it to be generic for future cases if we need it. =)

renatho commented on assets/admin/editor-wizard/style.scss:
```diff
@@ -0,0 +1,15 @@
+.sensei-editor-wizard {
+	& &__footer {
```
Do we really need the extra `& ` here in these classes?

aaronfc replied:
Not really. I am just scared about "specificity" bugs in CSS hahaha. Changed here b32f0cefe

renatho commented on assets/admin/editor-wizard/steps/lesson-details-step.js:
```diff
@@ -0,0 +1,32 @@
+ * @param {Object}   props.data
+ * @param {Function} props.setData
+ */
+const LessonDetailsStep = ( { data: wizardData, setData: setWizardData } ) => {
```
Is there a reason that only in 2 steps you changed the prop names from `data` to `wizardData`? I wonder if it would be nice to keep the same name in all steps.

That said, I think both names work for me `data` or `wizardData`.

alexsanford replied:
I kinda like the `wizardData` naming, personally. Makes it more explicit. I would vote for calling it `wizardData` and `setWizardData` everywhere 🤷‍♂️

aaronfc replied:
No special reason. I wanted to undo the change to use `data` and I forgot to change it here. Updated here bb1971b7e

aaronfc replied:
>I kinda like the wizardData naming, personally. Makes it more explicit. I would vote for calling it wizardData and setWizardData everywhere 🤷‍♂️

Oops Alex, I just renamed them back to `data` here bb1971b. But I didn't really have any strong opinion and neither renatho so I will update them back to `wizardData`. Would you rather have the property name set to `wizardData` or use the property renaming I had from `data` to `wizardData`?

alexsanford replied:
Meh. No strong opinion from me 🙂

renatho reviewed (commented):
It's getting very nice! Added a few comments (small details).

renatho commented on assets/admin/editor-wizard/wizard.js:
```diff
@@ -0,0 +1,53 @@
+ * @param {Array}    props.steps        Array with the steps that will be rendered.
+ * @param {Function} props.onCompletion Callback to call when wizard is completed.
+ */
+const Wizard = ( { steps, onCompletion } ) => {
```
Just one extra thing. I don't know if you're planning to write tests here, but at least for the `Wizard` component it would be interesting. 😊

aaronfc replied:
Good idea! I haven't done any JS tests yet, but this looks like a nice first time :)

aaronfc replied:
Done! (And thanks for your help!)

alexsanford commented on assets/admin/editor-wizard/wizard.js:
```diff
@@ -0,0 +1,53 @@
+			</div>
+			<div className={ 'sensei-editor-wizard__footer' }>
+				<div className={ 'sensei-editor-wizard__progress' }>
+					Step { currentStepNumber } of { steps.length }
```
I think we should display `currentStepNumber + 1`, right?

aaronfc replied:
Correct. Changed here: 119ea9691

Thanks!

aaronfc committed:
Simplify CSS

alexsanford reviewed (commented):
This is looking good! I like the architecture 🙂

aaronfc committed:
Rename variables back to data.

aaronfc committed:
Display correct step number.

aaronfc committed:
Add `wizard.test.js`.

fjorgemota commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -81,10 +88,24 @@ const EditorWizardModal = () => {
 	};
 
+	// TODO Implement different flows depending on post type 👇.
+	const steps = [ CourseDetailsStep, UpgradeStep, CoursePatternsStep ];
```
I really liked this.

Initially, I was thinking about implementing something using the component's `children` (or maybe even React Contexts) to get the steps correctly, but that would definitely lead to a more complex implementation without a clear benefit.

I was a little worried about passing data through from the step to the other components, but I feel like the `setData` should be quite enough. I really liked your approach here. :D

aaronfc replied:
Thanks for the good words @fjorgemota. I am happy that we both followed more or less the same steps. I initially tried with `children` and then started playing around passing the context but as you said... I couldn't see the benefit and it was more complex (at least for my understanding).

🙌

fjorgemota commented on assets/admin/editor-wizard/wizard.js:
```diff
@@ -0,0 +1,53 @@
+				<CurrentStep
+					data={ data }
+					setData={ setData }
+					onCompletion={ onCompletion }
```
Do we have any specific cases where the CurrentStep needs to call directly the `onCompletion` callback?

I feel like this might happen on the last step - specifically the "Course Patterns" step - but I guess that this step is the final one we could just pass the `goToNextStep` here instead, in the same way, that we pass that callback to the `CurrentStep.Actions` component. :thinking:

aaronfc replied:
Well, that's a good point! Probably `goToNextStep` might be enough. I like being more "explicit" with the `onCompletion` but sure it's not really relevant and we could go with `goToNextStep`.

I will update the comment in the sample "pattern" steps so that whoever implements them takes into account both approaches.

aaronfc committed:
Fix test and use correct `expect` method.

fjorgemota reviewed (commented):
Looks good to me. I left some comments but nothing waaay too important, just some questions and also some praise about the implementation.

While reading this PR, I really felt like we could use some `useEvent` in some places (here its [RFC](https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md)), but that's totally not needed here, just something I liked to think about (and that I find it worth to mention, hehe) =)

aaronfc committed:
Add more tests for the Wizard.

aaronfc committed:
Define different steps for different post types.

aaronfc committed:
Rename `data` and `setData` in sample steps to `wizard*`.

aaronfc committed:
Mention alternative approach for patterns sample step.

aaronfc committed:
Make steps update the modal title.

aaronfc committed:
Add some styling for wizard.

aaronfc commented:
👋 ,
applied some of the suggested changes in the review and also implemented some further things:
- Modal title updated from steps through Wizard's data – I am not very convinced about this because I think it could make mare sense for the title of the modal being part of the steps itself. But that would imply messing with the "close modal" button. Maybe we can review this when implementing the steps.
- Added some styling – nothing really final but something we can improve on.

I don't plan to add any further changes to this review. Any other change will be part of separate tasks.

aaronfc committed:
Fix tests after adding `onChange` to `Wizard`.

- Adding also a test for the `onChange` method.

fjorgemota commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -88,24 +91,41 @@ const EditorWizardModal = () => {
 
+	const updateModalTitle = ( wizardData ) => {
+		if (
+			wizardData.modalTitle !== undefined &&
+			wizardData.modalTitle !== modalTitle
```
Technically we don't need this comparison, because React should be able to not re-render everything [if the value set is the same as the previous one](https://reactjs.org/docs/hooks-reference.html#usestate): 

> If you update a State Hook to the same value as the current state, React will bail out without rendering the children or firing effects. (React uses the [Object.is comparison algorithm](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description).)

There's this little issue with this approach, tho, but I'm not sure if this is a huge problem:

> Note that React may still need to render that specific component again before bailing out. That shouldn’t be a concern because React won’t unnecessarily go “deeper” into the tree. If you’re doing expensive calculations while rendering, you can optimize them with useMemo.

So in theory we could simplify the `if` condition to this:

```suggestion
			wizardData.modalTitle
```

fjorgemota commented on assets/admin/editor-wizard/steps/lesson-details-step.js:
```diff
@@ -1,21 +1,23 @@
-		setData( { ...data, title: event.target.value } );
-	};
+const LessonDetailsStep = ( { data: wizardData, setData: setWizardData } ) => {
+	// Update modal title.
+	useEffect( () => {
+		setWizardData( { ...wizardData, modalTitle: 'Lesson Details Step' } );
+	}, [] );
```
I feel like this is sounding like some sort of boilerplate, honestly, even if it's too small.

What do you think about adding the title as part of the component "static" properties, in the same way you did with the "Actions" static property? This way, we could change this whole `useEffect` by a simple additional static property:

```LessonDetailsStep.Title = 'Lesson Details Step'```

And then you can update the modalTitle appropriately when the steps change :)

aaronfc replied:
I really like the simplicity on this approach.

Change 0257a7b25.

fjorgemota commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -88,24 +91,41 @@ const EditorWizardModal = () => {
-			`Wizard completed with data: ${ JSON.stringify( data ) }`
-		);
+	// Choose steps by post type.
+	const stepsByPostType = {
+		course: [ CourseDetailsStep, UpgradeStep, CoursePatternsStep ],
+		lesson: [ LessonDetailsStep, LessonPatternsStep ],
+	};
+	const postType = select( 'core/editor' )?.getCurrentPostType();
+	const steps = stepsByPostType[ postType ];
```
Nice solution to get the steps of the wizard by post type.

Maybe it would be worth to add a comment to backlink [to this file](https://github.com/Automattic/sensei/blob/086323830efcda4c10a2add2928b80a7a3cfdb87/includes/admin/class-sensei-editor-wizard.php#L83) (and maybe add a link there too) so we can avoid a situation where the editor wizard is included on a post type that is not supported. 

WDYT?

aaronfc replied:
Thanks for your comment! Instead of this I have updated the code to just ignore a non-supported post type.
In case we add a new post type and we want to show some wizard it should be pretty obvious how to find this and change it.

Change: 763a9e9

fjorgemota reviewed (commented):
Great changes!

Just left some additional comments about your latest changes. =)

renatho commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -81,10 +91,41 @@ const EditorWizardModal = () => {
+		course: [ CourseDetailsStep, UpgradeStep, CoursePatternsStep ],
+		lesson: [ LessonDetailsStep, LessonPatternsStep ],
+	};
+	const postType = select( 'core/editor' )?.getCurrentPostType();
```
I think since we're in a React scope, it would be interesting to use the `useSelect` here.

```js
const { postType } = useSelect( select => ( {
    postType: select( editorStore )?.getCurrentPostType(),
} );
```

Notice that it's also interesting to use the `editorStore` here since it's the new way to do it.

aaronfc replied:
Changed here aff5ae0e0

Thanks!

aaronfc committed:
Simplify modal title condition.

Co-authored-by: Fernando Jorge Mota <fjorgemota@users.noreply.github.com>

aaronfc committed:
Unsupported post type won't open modal.

renatho commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -81,10 +91,41 @@ const EditorWizardModal = () => {
+				<Wizard
+					steps={ steps }
+					onCompletion={ onWizardCompletion }
+					onChange={ updateModalTitle }
```
This seems a little confusing to me. The onChange in the `Wizard` and the `updateModalTitle` here seems something more generic. As you said in other comment, I'd remove this and use the title as part of the steps. Otherwise to style it would be a little confusing too since in some steps it's just in a left column, and in the patterns it would be full width.

aaronfc replied:
@renatho I have simplified all of this based on input from @fjorgemota here https://github.com/Automattic/sensei/commit/763a9e9ce637a4035adb8eaca5c9d3ecb4fcb128

I would rather let it as it is now and while implementing the first step we can give it a go styling the title inside the steps themselves. What do you think?

aaronfc committed:
Simplify modal title into static properties in Steps.

aaronfc committed:
Change `postType` calculation.

aaronfc committed:
Update with feature branch.

renatho reviewed (approved):
As talked about in the comments, I think probably it'll be better to remove the modal title logic and handle it as part of the content of the steps. But I think we can go ahead and revisit it later. 😉
